File: python_test/algorithm/graph_preprocessor.py
"""
Phase 1-2: Graph Preprocessor

Preprocesses raw relations:
- Deduplication (removes bidirectional duplicates)
- Entity ordering by connectivity
"""

from typing import List, Tuple
from collections import defaultdict


class GraphPreprocessor:
    """
    Preprocesses graph relations before layer classification
    """

    @staticmethod
    def deduplicate(relations: List[Tuple[str, str]]) -> List[Tuple[str, str]]:
        """
        Remove duplicate relations based on entity pairs (order-agnostic)

        This matches algo10's deduplication logic:
        - (A, B) and (A, B) -> keep one
        - (A, B) and (B, A) -> keep one (bidirectional treated as same)

        Example: users -> profiles AND profiles -> users
                 Only the first one is kept
        """
        seen_pairs = {}
        unique_relations = []

        for a, b in relations:
            pair_key = frozenset({a, b})  # Order-agnostic
            if pair_key not in seen_pairs:
                seen_pairs[pair_key] = (a, b)
                unique_relations.append((a, b))

        return unique_relations

    @staticmethod
    def build_entity_order(relations: List[Tuple[str, str]]) -> List[str]:
        """
        Build entity processing order based on connectivity

        Algorithm:
        1. Sort by connection count (descending)
        2. Breadth-first expansion from most connected
        3. Tie-breaker: position in connectivity ranking
        """
        # Count connections per entity
        connection_count = defaultdict(int)
        for a, b in relations:
            connection_count[a] += 1
            connection_count[b] += 1

        # Rule 1: Sort by connectivity
        connectivity_ranking = sorted(
            connection_count.keys(),
            key=lambda x: connection_count[x],
            reverse=True
        )

        # Start with most connected entity
        entity_order = []
        frontier = []

        if connectivity_ranking:
            entity_order.append(connectivity_ranking[0])

            # Initialize frontier with neighbors of first entity
            for a, b in relations:
                if a == entity_order[0] and b not in frontier:
                    frontier.append(b)
                if b == entity_order[0] and a not in frontier:
                    frontier.append(a)

        # Rule 2: Breadth-first expansion with connectivity tie-breaker
        while len(entity_order) < len(connectivity_ranking):
            candidates = [e for e in frontier if e not in entity_order]

            if not candidates:
                # No more in frontier, pick from unvisited
                candidates = [e for e in connectivity_ranking if e not in entity_order]
                if not candidates:
                    break

            # Rule 3: Tie-breaker by connectivity ranking
            next_entity = max(
                candidates,
                key=lambda e: (
                    connection_count[e],
                    -connectivity_ranking.index(e)
                )
            )

            entity_order.append(next_entity)

            # Expand frontier with neighbors
            for a, b in relations:
                if a == next_entity and b not in frontier and b not in entity_order:
                    frontier.append(b)
                if b == next_entity and a not in frontier and a not in entity_order:
                    frontier.append(a)

        return entity_order

File: src/domain/services/layout/GraphPreprocessor.ts
/**
 * Phase 1-2: Graph Preprocessor
 *
 * Preprocesses raw relations:
 * - Deduplication (removes bidirectional duplicates)
 * - Entity ordering by connectivity
 */

import { DirectedRelation } from './types';
import { Logger } from '../../../infrastructure/utils/Logger';

export interface PreprocessorResult {
  relations: DirectedRelation[];
  connectionCount: Map<string, number>;
}

export class GraphPreprocessor {
  /**
   * ÉTAPE 1: Deduplicate relations based on entity pairs (order-agnostic)
   *
   * Matches algo10's deduplication logic:
   * - (A, B) and (A, B) -> keep one
   * - (A, B) and (B, A) -> keep one (bidirectional treated as same)
   *
   * @param relationsRaw - Raw relations from parser
   * @returns Deduplicated relations and connection counts
   */
  static buildBacklog(relationsRaw: DirectedRelation[]): PreprocessorResult {
    Logger.subsection('PHASE 1: DEDUPLICATION (BUILD BACKLOG)');

    // Deduplication based on entity PAIR (order-agnostic)
    const seenPairs = new Map<string, DirectedRelation>();
    const uniqueRelations: DirectedRelation[] = [];
    const duplicatesRemoved: string[] = [];

    for (const { left, right } of relationsRaw) {
      // Order-agnostic key: alphabetically sorted pair
      const pairKey = [left, right].sort().join('|');

      if (!seenPairs.has(pairKey)) {
        // First time seeing this pair
        seenPairs.set(pairKey, { left, right });
        uniqueRelations.push({ left, right });
      } else {
        // Duplicate detected!
        const first = seenPairs.get(pairKey)!;
        duplicatesRemoved.push(`  [DUPLICATE] ${left} > ${right} (first: ${first.left} > ${first.right})`);
      }
    }

    Logger.debug(`After deduplication: ${uniqueRelations.length} unique relations`);

    if (duplicatesRemoved.length > 0) {
      Logger.debug(`\n${duplicatesRemoved.length} duplicate(s) removed:`);
      duplicatesRemoved.slice(0, 5).forEach(dup => Logger.debug(dup));
      if (duplicatesRemoved.length > 5) {
        Logger.debug(`  ... (${duplicatesRemoved.length - 5} more)`);
      }
    } else {
      Logger.debug('No duplicates detected');
    }

    // Count connections
    const connectionCount = new Map<string, number>();
    for (const { left, right } of uniqueRelations) {
      connectionCount.set(left, (connectionCount.get(left) || 0) + 1);
      connectionCount.set(right, (connectionCount.get(right) || 0) + 1);
    }

    return { relations: uniqueRelations, connectionCount };
  }

  /**
   * ÉTAPE 2: Build entity processing order based on connectivity
   *
   * Algorithm uses 3 criteria:
   * 1. Liste-Règle 1: sorted by connection count (descending)
   * 2. Connection with already chosen entities
   * 3. Tie-breaker: position in Liste-Règle 1
   *
   * @param relations - Deduplicated relations
   * @param connectionCount - Connection count per entity
   * @returns Ordered list of entities for processing
   */
  static buildEntityOrder(
    relations: DirectedRelation[],
    connectionCount: Map<string, number>
  ): string[] {
    Logger.subsection('PHASE 2: ENTITY ORDERING');

    // Liste-Règle 1 (reference)
    const listeRegle1 = Array.from(connectionCount.keys()).sort(
      (a, b) => connectionCount.get(b)! - connectionCount.get(a)!
    );

    // List to fill
    const entityOrder: string[] = [];
    const listeEnonces: string[] = [];

    // ITERATION 1: Take the first element (most connected)
    entityOrder.push(listeRegle1[0]);

    // Add connected entities
    for (const { left, right } of relations) {
      if (left === entityOrder[0] && !listeEnonces.includes(right)) {
        listeEnonces.push(right);
      }
      if (right === entityOrder[0] && !listeEnonces.includes(left)) {
        listeEnonces.push(left);
      }
    }

    // SUBSEQUENT ITERATIONS
    while (entityOrder.length < listeRegle1.length) {
      const candidates = listeEnonces.filter(e => !entityOrder.includes(e));
      if (candidates.length === 0) break;

      // Choose candidate with highest connection count
      // Tie-breaker: position in listeRegle1
      const nextEntity = candidates.reduce((best, current) => {
        const currentConnections = connectionCount.get(current) || 0;
        const bestConnections = connectionCount.get(best) || 0;

        if (currentConnections > bestConnections) return current;
        if (currentConnections < bestConnections) return best;

        // Tie-breaker: position in listeRegle1
        return listeRegle1.indexOf(current) < listeRegle1.indexOf(best) ? current : best;
      });

      entityOrder.push(nextEntity);

      // Update liste_enonces
      for (const { left, right } of relations) {
        if (left === nextEntity && !listeEnonces.includes(right) && !entityOrder.includes(right)) {
          listeEnonces.push(right);
        }
        if (right === nextEntity && !listeEnonces.includes(left) && !entityOrder.includes(left)) {
          listeEnonces.push(left);
        }
      }
    }

    Logger.debug(`Entity order (top 10): ${entityOrder.slice(0, 10).join(' > ')}`);
    if (entityOrder.length > 10) {
      Logger.debug(`  ... (${entityOrder.length - 10} more)`);
    }

    return entityOrder;
  }
}
